library/core/src/option.rs

```diff
@@ -1668,7 +1668,11 @@ const fn expect_failed(msg: &str) -> ! {
 /////////////////////////////////////////////////////////////////////////////
 
 #[stable(feature = "rust1", since = "1.0.0")]
-impl<T: Clone> Clone for Option<T> {
+#[rustc_const_unstable(feature = "const_clone", issue = "91805")]
+impl<T> const Clone for Option<T>
+where
+    T: ~const Clone + ~const Drop,
```

Comment:
Does this change the way rustdoc renders it?

Reply:
Before:
![image](https://user-images.githubusercontent.com/77098789/151511073-5456d4c5-bd2b-407e-99c2-cc79196f4dad.png)
After:
![image](https://user-images.githubusercontent.com/77098789/151510884-632dda94-3922-44a3-9dba-958f0f853ffa.png)

Reply:
ah yea, there's a pending fix for the missing space, but the additional `Drop` bound is really confusing for users and we should probably figure out out story there (both for documentation and explanation) before we expose it

Reply:
Will all such cases need to be explained in the future? It seems a bit hard... I think it would be nice if we could switch displaying of `~const` (and `~const Drop`) bound...

Reply:
Yea, I'm unsure about `~const` bounds in general, we should probably talk to the rustdoc team to figure out how it should display.

For `~const Drop` I think we should just hide it from rustdoc. That bound is pretty much useless to see for anyone. We could potentially add a small symbol to the generic param that has a `~const Drop` bound. Otherwise I fear that the useful bounds may get drowned out

Reply:
I agree with you that adding a small symbol is a good idea. IMO, `~const` bound itself is similar to optional bound, it's relatively easy to understand, so it may not impossible to be displayed as it is.

Reply:
As someone who doesn't work on `~const` but does work on rustdoc, I have to admit I'm unclear what the purpose of `~const Drop` is. It seems like nearly every API needs to add it, which adds a lot of clutter to the source code (not just the docs).

Reply:
`T: ~const Drop` seems to mean "T doesn't require non-const Drop to destruct itself", not "T implements `~const Drop`". For example, primitive types don't implement `const Drop` but are destructible at compile-time.

This judging seems to be implemented here:
https://github.com/rust-lang/rust/blob/master/compiler/rustc_const_eval/src/transform/check_consts/qualifs.rs#L140

Reply:
> `T: ~const Drop` seems to mean "T doesn't require non-const `Drop` to destruct the value"

I'm a bit confused about what you mean. Do you mean "`T` must be able to be `Drop`ed at compile-time"?

Reply:
Yes. I'm sorry for confusing you.

Reply:
What does the `~` mean? I thought `~const` meant "maybe const", but I may have misunderstood.

Reply:
It means "const if whatever the bound is on is in a const context". So if it's on a generic param of a const fn, calling that const fn in a const item or other const context requires the generic param to implement the trait constly, in regular runtime code it just requires the trait to be implemented with no requirement about constness at all